extensions-contrib/materialized-view-selection/src/main/java/io/druid/query/materializedview/DatasourceOptimizer.java

```diff
@@ -0,0 +1,212 @@
+        missFields.get(datasourceName).putIfAbsent(requiredFields, new AtomicLong(0));
+        missFields.get(datasourceName).get(requiredFields).incrementAndGet();
+        costTime.get(datasourceName).addAndGet(System.currentTimeMillis() - start);
+        return Lists.newArrayList(query);
+      }
+
+      // 
+      List<Query> queries = Lists.newArrayList();
+      List<Interval> remainingQueryIntervals = (List<Interval>) query.getIntervals();
+      for (Derivative derivative : derivativesWithRequiredFields) {
+        List<Interval> derivativeIntervals = Lists.newArrayList();
+        for (Interval interval : remainingQueryIntervals) {
+          serverView.getTimeline(new TableDataSource(derivative.getName()))
+              .lookup(interval)
+              .forEach(new Consumer<TimelineObjectHolder<String, ServerSelector>>() {
+                @Override
+                public void accept(TimelineObjectHolder<String, ServerSelector> stringServerSelectorTimelineObjectHolder) 
+                {
+                  derivativeIntervals.add(stringServerSelectorTimelineObjectHolder.getInterval());
+                }
+              }
+            );
+        }
+        // if the derivative does not contain any parts of intervals in the query, the derivative will not be selected. 
+        if (derivativeIntervals.isEmpty()) {
+          continue;
+        }
+        
+        remainingQueryIntervals = MaterializedViewUtils.minus(remainingQueryIntervals, derivativeIntervals);
+        queries.add(
+            query.withDataSource(new TableDataSource(derivative.getName()))
+                .withQuerySegmentSpec(new MultipleIntervalSegmentSpec(derivativeIntervals))
+        );
+        derivativesHitCount.get(derivative).incrementAndGet();
+        if (remainingQueryIntervals.isEmpty()) {
+          break;
+        }
+      }
+
+      if (queries.isEmpty()) {
+        costTime.get(datasourceName).addAndGet(System.currentTimeMillis() - start);
+        return Lists.newArrayList(query);
+      }
+
+      //after materialized view selection, the result of the remaining query interval will be computed based on the original datasource. 
+      if (!remainingQueryIntervals.isEmpty()) {
+        queries.add(query.withQuerySegmentSpec(new MultipleIntervalSegmentSpec(remainingQueryIntervals)));
+      }
+      hitCount.get(datasourceName).incrementAndGet();
+      costTime.get(datasourceName).addAndGet(System.currentTimeMillis() - start);
+      return queries;
+    } 
+    finally {
+      lock.readLock().unlock();
+    }
+  }
+
+  public List<DatasourceOptimizerStats> getAndResetStats() 
+  {
+    ImmutableMap<Derivative, AtomicLong> derivativesHitCountSnapshot;
+    ImmutableMap<String, AtomicLong> totalCountSnapshot;
+    ImmutableMap<String, AtomicLong> hitCountSnapshot;
+    ImmutableMap<String, AtomicLong> costTimeSnapshot;
+    ImmutableMap<String, ConcurrentHashMap<Set<String>, AtomicLong>> missFieldsSnapshot;
+    try {
```

Comment:
This should be outside of the `try` clause (https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/Lock.html).